File: examples/crm/crm.c
#include "TACSMeshLoader.h"
#include "MITCShell.h"
#include "isoFSDTStiffness.h"

int main( int argc, char **argv ){
  // Intialize MPI and declare communicator
  MPI_Init(&argc, &argv);
  MPI_Comm comm = MPI_COMM_WORLD;

  // Write name of BDF file to be load to char array
  const char *filename = "CRM_box_2nd.bdf";

  // Create the mesh loader object and load file
  TACSMeshLoader *mesh = new TACSMeshLoader(comm);
  mesh->incref();
  mesh->scanBDFFile(filename);

  // Get number of components prescribed in BDF file
  int num_components = mesh->getNumComponents();

  // Set properties needed to create stiffness object 
  double rho = 2500.0; // density, kg/m^3
  double E = 70e9; // elastic modulus, Pa
  double nu = 0.3; // poisson's ratio
  double kcorr = 5.0/6.0; // shear correction factor
  double ys = 350e6; // yield stress, Pa

  // Loop over components, creating constituitive object for each
  for ( int i = 0; i < num_components; i++ ){
    const char *descriptor = mesh->getElementDescript(i);
    double min_thickness = 0.01;
    double max_thickness = 0.20;
    double thickness = 0.07;
    isoFSDTStiffness *stiff = new isoFSDTStiffness(rho, E, nu, kcorr, ys,
        thickness, i, min_thickness, max_thickness); 

    // Initialize element object
    TACSElement *element = NULL;

    // Create element object using constituitive information and type defined in
    // descriptor
    if ( strcmp(descriptor, "CQUAD") == 0 || 
         strcmp(descriptor, "CQUADR") == 0 ||
         strcmp(descriptor, "CQUAD4") == 0) {
      element = new MITCShell<2>(stiff, LINEAR, i);
    }
    mesh->setElement(i, element);
  }
  
  // Create tacs assembler from mesh loader object
  TACSAssembler *tacs = mesh->createTACS(6);
  tacs->incref();
  mesh->decref();

#ifdef TACS_USE_COMPLEX
  // Get the deseign variable values
  TacsScalar *x = new TacsScalar[ num_components ];
  memset(x, 0, num_components*sizeof(TacsScalar));

  // Get the design variable values
  tacs->getDesignVars(x, num_components);

  double dh = 1e-30;
  for ( int i = 0; i < num_components; i++ ){
    x[i] = x[i] + TacsScalar(0.0, dh);
  }

  // Set the perturbed design variable values
  tacs->setDesignVars(x, num_components);
#endif // TACS_USE_COMPLEX

  // Create matrix and vectors 
  TACSBVec *ans = tacs->createVec(); // displacements and rotations
  TACSBVec *f = tacs->createVec(); // loads
  FEMat *mat = tacs->createFEMat(); // preconditioner

  // Increment reference count to the matrix/vectors
  ans->incref();
  f->incref();
  mat->incref();

  // Allocate the factorization
  int lev = 10000;
  double fill = 10.0;
  int reorder_schur = 1;
  PcScMat *pc = new PcScMat(mat, lev, fill, reorder_schur); 
  pc->incref();

  // Assemble and factor the stiffness/Jacobian matrix
  double alpha = 1.0, beta = 0.0, gamma = 0.0;
  tacs->assembleJacobian(alpha, beta, gamma, NULL, mat);
  mat->applyBCs();
  pc->factor(); // LU factorization of stiffness matrix

  // Set all the entries in load vector to specified value
  f->set(1.0);
  f->applyBCs();

  // Get solution and store in ans
  pc->applyFactor(f, ans);
  tacs->setVariables(ans);

  // Print out the norm of the solution on the root processor
  TacsScalar norm = ans->norm();
  int rank;
  MPI_Comm_rank(comm, &rank);

  if (rank == 0){
#ifdef TACS_USE_COMPLEX
    printf("||ans||: %e + j%e\n", RealPart(norm), ImagPart(norm));
#else
    printf("||ans||: %e\n", ans->norm());
#endif
  }

  // Create an TACSToFH5 object for writing output to files
  unsigned int write_flag = (TACSElement::OUTPUT_NODES |
                             TACSElement::OUTPUT_DISPLACEMENTS |
                             TACSElement::OUTPUT_STRAINS |
                             TACSElement::OUTPUT_STRESSES |
                             TACSElement::OUTPUT_EXTRAS);
  TACSToFH5 * f5 = new TACSToFH5(tacs, SHELL, write_flag);
  f5->incref();

  // Write the displacements
  f5->writeToFile("ucrm.f5");

  // Delete the viewer
  f5->decref();

  mat->decref();
  pc->decref();
  tacs->decref();

  MPI_Finalize();

  return 0;
}

File: examples/crm/crm.py
# A demonstration of basic functions of the Python interface for TACS,
# this example goes through the process of setting up the using the
# tacs assembler directly as opposed to using the pyTACS user interface (see analysis.py):
# loading a mesh, creating elements, evaluating functions, solution, and output
from __future__ import print_function

# Import necessary libraries
import numpy as np
import os
from mpi4py import MPI
from tacs import TACS, elements, constitutive, functions

# Load structural mesh from BDF file
bdfFile = os.path.join(os.path.dirname(__file__), "CRM_box_2nd.bdf")
tacs_comm = MPI.COMM_WORLD
struct_mesh = TACS.MeshLoader(tacs_comm)
struct_mesh.scanBDFFile(bdfFile)

# Set constitutive properties
rho = 2500.0  # density, kg/m^3
E = 70e9  # elastic modulus, Pa
nu = 0.3  # poisson's ratio
kcorr = 5.0 / 6.0  # shear correction factor
ys = 350e6  # yield stress, Pa
min_thickness = 0.002
max_thickness = 0.20
thickness = 0.02

# Loop over components, creating stiffness and element object for each
num_components = struct_mesh.getNumComponents()
for i in range(num_components):
    descriptor = struct_mesh.getElementDescript(i)
    # Setup (isotropic) property and constitutive objects
    prop = constitutive.MaterialProperties(rho=rho, E=E, nu=nu, ys=ys)
    # Set one thickness dv for every component
    stiff = constitutive.IsoShellConstitutive(
        prop, t=thickness, tMin=min_thickness, tMax=max_thickness, tNum=i
    )

    element = None
    transform = None
    if descriptor in ["CQUAD", "CQUADR", "CQUAD4"]:
        element = elements.Quad4Shell(transform, stiff)
    struct_mesh.setElement(i, element)

# Create tacs assembler object from mesh loader
tacs = struct_mesh.createTACS(6)

# Create the KS Function
ksWeight = 100.0
funcs = [functions.KSFailure(tacs, ksWeight=ksWeight)]
# funcs = [functions.StructuralMass(tacs)]
# funcs = [functions.Compliance(tacs)]

# Get the design variable values
x = tacs.createDesignVec()
x_array = x.getArray()
tacs.getDesignVars(x)

# Get the node locations
X = tacs.createNodeVec()
tacs.getNodes(X)
tacs.setNodes(X)

# Create the forces
forces = tacs.createVec()
force_array = forces.getArray()
force_array[2::6] += 100.0  # uniform load in z direction
tacs.applyBCs(forces)

# Set up and solve the analysis problem
res = tacs.createVec()
ans = tacs.createVec()
u = tacs.createVec()
mat = tacs.createSchurMat()
pc = TACS.Pc(mat)
subspace = 100
restarts = 2
gmres = TACS.KSM(mat, pc, subspace, restarts)

# Assemble the Jacobian and factor
alpha = 1.0
beta = 0.0
gamma = 0.0
tacs.zeroVariables()
tacs.assembleJacobian(alpha, beta, gamma, res, mat)
pc.factor()

# Solve the linear system
gmres.solve(forces, ans)
tacs.setVariables(ans)

# Evaluate the function
fvals1 = tacs.evalFunctions(funcs)

# Solve for the adjoint variables
adjoint = tacs.createVec()
res.zeroEntries()
tacs.addSVSens([funcs[0]], [res])
gmres.solve(res, adjoint)

# Compute the total derivative w.r.t. material design variables
fdv_sens = tacs.createDesignVec()
fdv_sens_array = fdv_sens.getArray()
tacs.addDVSens([funcs[0]], [fdv_sens])
tacs.addAdjointResProducts([adjoint], [fdv_sens], -1)
# Finalize sensitivity arrays across all procs
fdv_sens.beginSetValues()
fdv_sens.endSetValues()

# Create a random direction along which to perturb the nodes
pert = tacs.createNodeVec()
X_array = X.getArray()
pert_array = pert.getArray()
pert_array[0::3] = X_array[1::3]
pert_array[1::3] = X_array[0::3]
pert_array[2::3] = X_array[2::3]

# Compute the total derivative w.r.t. nodal locations
fXptSens = tacs.createNodeVec()
tacs.addXptSens([funcs[0]], [fXptSens])
tacs.addAdjointResXptSensProducts([adjoint], [fXptSens], -1)
# Finalize sensitivity arrays across all procs
fXptSens.beginSetValues()
fXptSens.endSetValues()

# Set the complex step
xpert = tacs.createDesignVec()
xpert.setRand()
xpert_array = xpert.getArray()
xnew = tacs.createDesignVec()
xnew.copyValues(x)
if TACS.dtype is complex:
    dh = 1e-30
    xnew.axpy(dh * 1j, xpert)
else:
    dh = 1e-6
    xnew.axpy(dh, xpert)

# Set the design variables
tacs.setDesignVars(xnew)

# Compute the perturbed solution
tacs.zeroVariables()
tacs.assembleJacobian(alpha, beta, gamma, res, mat)
pc.factor()
gmres.solve(forces, u)
tacs.setVariables(u)

# Evaluate the function for perturbed solution
fvals2 = tacs.evalFunctions(funcs)

if TACS.dtype is complex:
    fd = fvals2.imag / dh
else:
    fd = (fvals2 - fvals1) / dh

result = xpert.dot(fdv_sens)
if tacs_comm.rank == 0:
    print("FD:      ", fd[0])
    print("Result:  ", result)
    print("Rel err: ", (result - fd[0]) / result)

# Reset the old variable values
tacs.setDesignVars(x)

if TACS.dtype is complex:
    dh = 1e-30
    X.axpy(dh * 1j, pert)
else:
    dh = 1e-6
    X.axpy(dh, pert)

# Set the perturbed node locations
tacs.setNodes(X)

# Compute the perturbed solution
tacs.zeroVariables()
tacs.assembleJacobian(alpha, beta, gamma, res, mat)
pc.factor()
gmres.solve(forces, u)
tacs.setVariables(u)

# Evaluate the function again
fvals2 = tacs.evalFunctions(funcs)

if TACS.dtype is complex:
    fd = fvals2.imag / dh
else:
    fd = (fvals2 - fvals1) / dh

# Compute the projected derivative
result = pert.dot(fXptSens)

if tacs_comm.rank == 0:
    print("FD:      ", fd[0])
    print("Result:  ", result)
    print("Rel err: ", (result - fd[0]) / result)

# Output for visualization
flag = (
    TACS.OUTPUT_CONNECTIVITY
    | TACS.OUTPUT_NODES
    | TACS.OUTPUT_DISPLACEMENTS
    | TACS.OUTPUT_STRAINS
    | TACS.OUTPUT_STRESSES
    | TACS.OUTPUT_EXTRAS
)
f5 = TACS.ToFH5(tacs, TACS.BEAM_OR_SHELL_ELEMENT, flag)
f5.writeToFile("ucrm.f5")
